Modules/_json.c

```diff
@@ -1863,40 +1863,48 @@ static PyMethodDef speedups_methods[] = {
 PyDoc_STRVAR(module_doc,
 "json speedups\n");
 
-static struct PyModuleDef jsonmodule = {
-        PyModuleDef_HEAD_INIT,
-        "_json",
-        module_doc,
-        -1,
-        speedups_methods,
-        NULL,
-        NULL,
-        NULL,
-        NULL
-};
-
-PyMODINIT_FUNC
-PyInit__json(void)
+static int
+_json_exec(PyObject *module)
 {
-    PyObject *m = PyModule_Create(&jsonmodule);
-    if (!m)
-        return NULL;
     if (PyType_Ready(&PyScannerType) < 0)
         goto fail;
     if (PyType_Ready(&PyEncoderType) < 0)
         goto fail;
     Py_INCREF((PyObject*)&PyScannerType);
-    if (PyModule_AddObject(m, "make_scanner", (PyObject*)&PyScannerType) < 0) {
+    if (PyModule_AddObject(module, "make_scanner", (PyObject*)&PyScannerType) < 0) {
         Py_DECREF((PyObject*)&PyScannerType);
         goto fail;
     }
     Py_INCREF((PyObject*)&PyEncoderType);
-    if (PyModule_AddObject(m, "make_encoder", (PyObject*)&PyEncoderType) < 0) {
+    if (PyModule_AddObject(module, "make_encoder", (PyObject*)&PyEncoderType) < 0) {
         Py_DECREF((PyObject*)&PyEncoderType);
         goto fail;
     }
-    return m;
+    return 0;
   fail:
-    Py_DECREF(m);
-    return NULL;
+    Py_DECREF(module);
```

Comment:
DECREF on error looks like a bug. For example, arraymodule.c doesn't do that. I suggest to remove the "fail" label and replace "goto fail" with "return -1".

Reply:
Good catch, thanks.